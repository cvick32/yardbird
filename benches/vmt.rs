use std::path::Path;
use std::time::Duration;

use criterion::{criterion_group, criterion_main, Criterion, SamplingMode};
use smt2parser::vmt::VMTModel;
use yardbird::{strategies::Abstract, Driver};

fn run(path: impl AsRef<Path>, depth: u16) {
    let vmt_model = VMTModel::from_path(path).unwrap();

    let cfg = z3::Config::new();
    let context = z3::Context::new(&cfg);
    let mut driver = Driver::new(&context, vmt_model);

    let strat = Box::new(Abstract::new(depth, false));

    _ = driver.check_strategy(depth, strat);
}

fn bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("vmt");
    group.sampling_mode(SamplingMode::Flat);
    group.bench_function("array_double_inverse", |b| {
        b.iter(|| run("examples/array_double_inverse.vmt", 10))
    });
    // group.bench_function("array_doub_access_init_const", |b| {
    //     b.iter(|| run("examples/array_doub_access_init_const.vmt", 10))
    // });
    group.sampling_mode(SamplingMode::Flat);
    group.finish();
}

criterion_group! {
    name = benches;
    config = Criterion::default()
        .sample_size(10)
        .warm_up_time(Duration::from_secs(50))
        .measurement_time(Duration::from_secs(200));
    targets = bench,
}
criterion_main!(benches);
