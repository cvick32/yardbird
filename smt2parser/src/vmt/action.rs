use crate::concrete::Command;



#[derive(Clone, Debug)]
pub struct Action {
    pub action: Command,
    pub relationship: Command,
}

impl Action {
    pub fn get_current_action_name(&self) -> &String {
        match &self.action {
            Command::DeclareFun {
                symbol,
                parameters: _,
                sort: _,
            } => {
                return &symbol.0;
            }
            _ => panic!("Actions's Command must be DeclareFun."),
        }
    }
    
    pub(crate) fn as_commands(&self) -> Vec<Command> {
        //(define-fun .grantExclusiveRule () Bool (! grantExclusiveRule :action 0))
        vec![self.action.clone(), self.relationship.clone()]
    }
}