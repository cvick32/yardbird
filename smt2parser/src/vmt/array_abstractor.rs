use std::collections::HashMap;

use num::{BigUint, Zero};

use crate::concrete::{Command, Identifier, QualIdentifier, Sort, Symbol, SyntaxBuilder, Term};

/// Rewrites Commands to use uninterpreted functions instead of explicit Arrays.
/// Currently turns all Arrays into Arr-Int-Int, but will be extended to arbitrary
/// Arrays later.
#[derive(Clone)]
pub struct ArrayAbstractor {
    pub visitor: SyntaxBuilder,
    pub array_types: HashMap<String, String>,
}
impl ArrayAbstractor {
    pub(crate) fn get_array_type_definitions(&self) -> Vec<Command> {
        let mut commands = vec![];
        for (index, value) in &self.array_types {
            let arr_sort = Sort::Simple {
                identifier: Identifier::Simple {
                    symbol: Symbol(format!("Array-{}-{}", index, value)),
                },
            };
            let index_sort = Sort::Simple {
                identifier: Identifier::Simple {
                    symbol: Symbol(index.to_string()),
                },
            };
            let value_sort = Sort::Simple {
                identifier: Identifier::Simple {
                    symbol: Symbol(value.to_string()),
                },
            };
            let sort_definition = Command::DeclareSort {
                symbol: Symbol(format!("Array-{}-{}", index, value)),
                arity: BigUint::zero(),
            };
            let read_definition: Command = Command::DeclareFun {
                symbol: Symbol(format!("Read-{}-{}", index, value)),
                parameters: vec![arr_sort.clone(), index_sort.clone()],
                sort: value_sort.clone(),
            };
            let write_definition: Command = Command::DeclareFun {
                symbol: Symbol(format!("Write-{}-{}", index, value)),
                parameters: vec![arr_sort.clone(), index_sort.clone(), value_sort.clone()],
                sort: arr_sort.clone(),
            };
            let constarr_definition: Command = Command::DeclareFun {
                symbol: Symbol(format!("ConstArr-{}-{}", index, value)),
                parameters: vec![value_sort],
                sort: arr_sort,
            };
            commands.push(sort_definition);
            commands.push(constarr_definition);
            commands.push(read_definition);
            commands.push(write_definition);
        }
        commands
    }
}

impl crate::rewriter::Rewriter for ArrayAbstractor {
    type V = SyntaxBuilder;
    type Error = crate::concrete::Error;

    fn visitor(&mut self) -> &mut Self::V {
        &mut self.visitor
    }

    fn visit_declare_fun(
        &mut self,
        symbol: <Self::V as crate::visitors::Smt2Visitor>::Symbol,
        parameters: Vec<<Self::V as crate::visitors::Smt2Visitor>::Sort>,
        sort: <Self::V as crate::visitors::Smt2Visitor>::Sort,
    ) -> Result<<Self::V as crate::visitors::Smt2Visitor>::Command, Self::Error> {
        let new_sort = match sort.clone() {
            crate::concrete::Sort::Simple { identifier: _ } => sort,
            crate::concrete::Sort::Parameterized {
                identifier,
                parameters: _,
            } => {
                if identifier.to_string() == "Array" {
                    crate::concrete::Sort::Simple {
                        identifier: Identifier::Simple {
                            symbol: Symbol("Array-Int-Int".to_string()),
                        },
                    }
                } else {
                    sort
                }
            }
        };
        Ok(Command::DeclareFun {
            symbol,
            parameters,
            sort: new_sort,
        })
    }

    fn visit_application(
        &mut self,
        qual_identifier: <Self::V as crate::visitors::Smt2Visitor>::QualIdentifier,
        arguments: Vec<<Self::V as crate::visitors::Smt2Visitor>::Term>,
    ) -> Result<<Self::V as crate::visitors::Smt2Visitor>::Term, Self::Error> {
        let new_identifier = match qual_identifier.clone() {
            crate::concrete::QualIdentifier::Simple { identifier } => {
                if identifier.to_string() == "select" {
                    simple_identifier_with_name("Read-Int-Int")
                } else if identifier.to_string() == "store" {
                    simple_identifier_with_name("Write-Int-Int")
                } else {
                    qual_identifier
                }
            }
            crate::concrete::QualIdentifier::Sorted {
                identifier,
                sort: _,
            } => {
                if identifier.to_string() == "const" {
                    simple_identifier_with_name("ConstArr-Int-Int")
                } else {
                    qual_identifier
                }
            }
        };
        Ok(Term::Application {
            qual_identifier: new_identifier,
            arguments,
        })
    }
}

fn simple_identifier_with_name(name: &str) -> QualIdentifier {
    crate::concrete::QualIdentifier::Simple {
        identifier: Identifier::Simple {
            symbol: Symbol(name.to_string()),
        },
    }
}
