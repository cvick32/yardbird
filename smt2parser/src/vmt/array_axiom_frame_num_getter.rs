use std::collections::BTreeSet;

use log::debug;

use crate::{
    concrete::{Symbol, SyntaxBuilder, Term},
    vmt::{variable::var_is_immutable, VARIABLE_FRAME_DELIMITER},
};

use super::variable::Variable;

#[derive(Clone, Debug)]
pub struct ArrayAxiomFrameNumGetter {
    pub visitor: SyntaxBuilder,
    pub instance_term: Term,
    pub array_term_frame_map: BTreeSet<(String, u64)>,
    pub int_term_frame_map: BTreeSet<(String, u64)>,
    variables: Vec<Variable>,
}

/// What happens on this example?
/// (not (= i@4 i@2)) => (Read-Int-Int (Write-Int-Int a@4 i@4 i@4) i@2) = (Read-Int-Int a@4 i@2)
/// We set i@2 = i and then quantify out everything else. I feel like this isn't
/// really what we want, what we want is to say forall i@2 then the rest holds. We don't
/// want to quantify over arrays. It's unclear to me if that even makes sense
impl ArrayAxiomFrameNumGetter {
    pub fn new(instance_term: Term, variables: Vec<Variable>) -> Self {
        let mut frame_getter = ArrayAxiomFrameNumGetter {
            visitor: SyntaxBuilder,
            instance_term: instance_term.clone(),
            array_term_frame_map: BTreeSet::new(),
            int_term_frame_map: BTreeSet::new(),
            variables,
        };

        instance_term.accept(&mut frame_getter).unwrap();

        frame_getter
    }

    pub fn max_array(&self) -> u64 {
        *self
            .array_term_frame_map
            .iter()
            .map(|(_, frame)| frame)
            .max()
            .unwrap_or(&0) // If all variables are immutable, return 0.
    }

    pub fn min_array(&self) -> u64 {
        *self
            .array_term_frame_map
            .iter()
            .map(|(_, frame)| frame)
            .min()
            .unwrap_or(&0) // If all variables are immutable, return 0.
    }

    pub fn max_int(&self) -> u64 {
        *self
            .int_term_frame_map
            .iter()
            .map(|(_, frame)| frame)
            .max()
            .unwrap_or(&0) // If all variables are immutable, return 0.
    }

    pub fn min_int(&self) -> u64 {
        *self
            .int_term_frame_map
            .iter()
            .map(|(_, frame)| frame)
            .min()
            .unwrap_or(&0) // If all variables are immutable, return 0.
    }

    fn get_var_sort(&self, var_name: &str) -> String {
        for variable in &self.variables {
            if variable.get_current_variable_name() == var_name {
                return variable.get_sort_name();
            }
        }
        panic!("Could not find variable {var_name} in {:?}", self.variables);
    }

    #[allow(clippy::type_complexity)]
    pub(crate) fn to_substitution(
        &self,
    ) -> Option<(
        std::collections::BTreeMap<(String, u64), String>,
        BTreeSet<String>,
        bool,
    )> {
        if self.max_array() - self.min_array() > 1 {
            // This forces us to never quantify over arrays.
            debug!(
                "[smt2parser] Tried to quantify over array in instantitation: {}",
                self.instance_term
            );
            None
        } else {
            let min_array_frame_number = self.min_array();
            let mut quantified = BTreeSet::new();
            let mut is_current = true;
            let mut subst: std::collections::BTreeMap<(String, u64), String> = self
                .int_term_frame_map
                .iter()
                .enumerate()
                .map(|(idx, (var, frame))| {
                    if *frame == min_array_frame_number || var_is_immutable(var)
                    // || (*frame == min_array_frame_number - 1
                    //     && self.max_array() == self.min_array())
                    {
                        ((var.clone(), *frame), var.clone())
                    } else if *frame == min_array_frame_number + 1 {
                        is_current = false;
                        ((var.clone(), *frame), format!("{var}_next"))
                    } else {
                        let name = format!("PH{idx}");
                        quantified.insert(name.clone());
                        ((var.clone(), *frame), name)
                    }
                })
                .collect();

            let arr_subst: std::collections::BTreeMap<(String, u64), String> = self
                .array_term_frame_map
                .iter()
                .map(|(var, frame)| {
                    if *frame == min_array_frame_number || var_is_immutable(var) {
                        ((var.clone(), *frame), var.clone())
                    } else {
                        is_current = false;
                        ((var.clone(), *frame), format!("{var}_next"))
                    }
                })
                .collect();

            subst.extend(arr_subst);
            Some((subst, quantified, is_current))
        }
    }
}

impl crate::rewriter::Rewriter for ArrayAxiomFrameNumGetter {
    type V = SyntaxBuilder;
    type Error = crate::concrete::Error;

    fn visitor(&mut self) -> &mut Self::V {
        &mut self.visitor
    }

    fn process_symbol(&mut self, s: Symbol) -> Result<Symbol, Self::Error> {
        if let Some((var_name, time_str)) = s.0.split_once(VARIABLE_FRAME_DELIMITER) {
            if !var_is_immutable(var_name) {
                let time = time_str.parse().unwrap();
                let var_sort = self.get_var_sort(var_name);
                if var_sort.contains("Array") {
                    self.array_term_frame_map
                        .insert((var_name.to_string(), time));
                } else {
                    self.int_term_frame_map.insert((var_name.to_string(), time));
                }
            }
        }
        Ok(s)
    }
}

/// Provides offset information for each variable in a term.
/// This is useful for understanding how to properly instantiate formulas during unrolling.
#[derive(Clone, Debug)]
pub struct VariableOffsetGetter {
    pub visitor: SyntaxBuilder,
    pub instance_term: Term,
    pub variable_offsets: std::collections::BTreeMap<String, Vec<i64>>,
    variables: Vec<Variable>,
}

impl VariableOffsetGetter {
    pub fn new(instance_term: Term, variables: Vec<Variable>) -> Self {
        let mut offset_getter = VariableOffsetGetter {
            visitor: SyntaxBuilder,
            instance_term: instance_term.clone(),
            variable_offsets: std::collections::BTreeMap::new(),
            variables,
        };

        instance_term.accept(&mut offset_getter).unwrap();

        offset_getter
    }

    /// Get the offset for a specific variable (returns the max offset for compatibility)
    pub fn get_offset(&self, var_name: &str) -> Option<i64> {
        self.variable_offsets
            .get(var_name)
            .and_then(|v| v.iter().max().copied())
    }

    /// Get all variable offsets
    pub fn get_all_offsets(&self) -> &std::collections::BTreeMap<String, Vec<i64>> {
        &self.variable_offsets
    }

    /// Get the minimum offset across all variables and all their offsets
    pub fn min_offset(&self) -> i64 {
        self.variable_offsets
            .values()
            .flat_map(|v| v.iter())
            .min()
            .copied()
            .unwrap_or(0)
    }

    /// Get the maximum offset across all variables and all their offsets
    pub fn max_offset(&self) -> i64 {
        self.variable_offsets
            .values()
            .flat_map(|v| v.iter())
            .max()
            .copied()
            .unwrap_or(0)
    }

    /// Get the total span of offsets (max - min)
    pub fn offset_span(&self) -> i64 {
        self.max_offset() - self.min_offset()
    }

    /// Check if all variables have the same offset (span = 0)
    pub fn is_uniform_offset(&self) -> bool {
        self.offset_span() == 0
    }

    fn get_var_sort(&self, var_name: &str) -> String {
        for variable in &self.variables {
            if variable.get_current_variable_name() == var_name {
                return variable.get_sort_name();
            }
        }
        panic!("Could not find variable {var_name} in {:?}", self.variables);
    }
}

impl crate::rewriter::Rewriter for VariableOffsetGetter {
    type V = SyntaxBuilder;
    type Error = crate::concrete::Error;

    fn visitor(&mut self) -> &mut Self::V {
        &mut self.visitor
    }

    fn process_symbol(&mut self, s: Symbol) -> Result<Symbol, Self::Error> {
        if let Some((var_name, time_str)) = s.0.split_once(VARIABLE_FRAME_DELIMITER) {
            if !var_is_immutable(var_name) {
                let time: u64 = time_str.parse().unwrap();
                // Calculate offset relative to the current frame (0)
                let offset = time as i64;
                self.variable_offsets
                    .entry(var_name.to_string())
                    .or_default()
                    .push(offset);
            }
        }
        Ok(s)
    }
}
