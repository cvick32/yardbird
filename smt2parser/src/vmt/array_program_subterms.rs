use std::{collections::HashSet, fmt::Debug};

use crate::{
    concrete::{Constant, Keyword, QualIdentifier, SExpr, Sort, Symbol, Term},
    visitors::TermVisitor,
    Error,
};

use super::utils::is_boolean_connective;

#[derive(Clone, Debug, Default)]
pub struct ArrayProgramSubterms {
    pub subterms: HashSet<Term>,
}

impl TermVisitor<Constant, QualIdentifier, Keyword, SExpr, Symbol, Sort> for ArrayProgramSubterms {
    type T = Term;
    type E = Error;

    /// Add contstant term to subterms.
    fn visit_constant(&mut self, constant: Constant) -> Result<Self::T, Self::E> {
        self.subterms.insert(Term::Constant(constant.clone()));
        Ok(Term::Constant(constant))
    }

    fn visit_qual_identifier(
        &mut self,
        qual_identifier: QualIdentifier,
    ) -> Result<Self::T, Self::E> {
        let qi = Term::QualIdentifier(qual_identifier.clone());
        if !is_boolean_connective(&qual_identifier) && !is_array_function_name(&qual_identifier) {
            self.subterms.insert(qi.clone());
        }
        Ok(qi)
    }

    /// Add non-boolean connective function applications.
    fn visit_application(
        &mut self,
        qual_identifier: QualIdentifier,
        arguments: Vec<Self::T>,
    ) -> Result<Self::T, Self::E> {
        for argument in &arguments {
            let _ = argument.clone().accept_term_visitor(self);
        }
        let app = Term::Application {
            qual_identifier: qual_identifier.clone(),
            arguments,
        };
        if !is_boolean_connective(&qual_identifier) {
            self.subterms.insert(app.clone());
        }
        Ok(app)
    }

    fn visit_let(
        &mut self,
        var_bindings: Vec<(Symbol, Self::T)>,
        term: Self::T,
    ) -> Result<Self::T, Self::E> {
        let _ = term.clone().accept_term_visitor(self);
        Ok(Term::Let {
            var_bindings,
            term: Box::new(term),
        })
    }

    fn visit_forall(
        &mut self,
        vars: Vec<(Symbol, Sort)>,
        term: Self::T,
    ) -> Result<Self::T, Self::E> {
        let _ = term.clone().accept_term_visitor(self);
        Ok(Term::Forall {
            vars,
            term: Box::new(term),
        })
    }

    fn visit_exists(
        &mut self,
        vars: Vec<(Symbol, Sort)>,
        term: Self::T,
    ) -> Result<Self::T, Self::E> {
        let _ = term.clone().accept_term_visitor(self);
        Ok(Term::Exists {
            vars,
            term: Box::new(term),
        })
    }

    fn visit_match(
        &mut self,
        term: Self::T,
        cases: Vec<(Vec<Symbol>, Self::T)>,
    ) -> Result<Self::T, Self::E> {
        Ok(Term::Match {
            term: Box::new(term),
            cases,
        })
    }

    fn visit_attributes(
        &mut self,
        term: Self::T,
        attributes: Vec<(
            Keyword,
            crate::concrete::AttributeValue<Constant, Symbol, SExpr>,
        )>,
    ) -> Result<Self::T, Self::E> {
        let _ = term.clone().accept_term_visitor(self);
        Ok(Term::Attributes {
            term: Box::new(term),
            attributes,
        })
    }
}

fn is_array_function_name(qual_identifier: &QualIdentifier) -> bool {
    let name = qual_identifier.get_name();
    name.contains("Read") || name.contains("Write") || name.contains("ConstArr")
}
