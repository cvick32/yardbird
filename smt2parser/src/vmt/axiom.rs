use crate::concrete::{Command, FunctionDec, Identifier, Sort, Symbol, Term};

#[derive(Clone, Debug)]
pub struct Axiom {
    pub _axiom: Term,
}

impl Axiom {
    pub(crate) fn _as_commands(&self) -> Vec<Command> {
        let dd = FunctionDec {
            name: Symbol("axiom".into()),
            parameters: vec![],
            result: Sort::Simple {
                identifier: Identifier::Simple {
                    symbol: Symbol("Bool".into()),
                },
            },
        };
        let command = Command::DefineFun {
            sig: dd,
            term: self._axiom.clone(),
        };
        vec![command]
    }
}
