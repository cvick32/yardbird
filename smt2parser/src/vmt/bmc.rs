use std::collections::HashMap;

use crate::concrete::{Symbol, SyntaxBuilder};

#[derive(Clone)]
pub struct BMCBuilder {
    pub visitor: SyntaxBuilder,
    pub current_variables: Vec<String>,
    pub next_variables: HashMap<String, String>,
    pub step: u8,
}

impl BMCBuilder {
    pub fn add_step(&mut self) {
        self.step += 1;
    }
}

impl crate::rewriter::Rewriter for BMCBuilder {
    type V = SyntaxBuilder;
    type Error = crate::concrete::Error;

    fn visitor(&mut self) -> &mut Self::V {
        &mut self.visitor
    }

    fn process_symbol(&mut self, s: Symbol) -> Result<Symbol, Self::Error> {
        if self.current_variables.contains(&s.0) {
            Ok(Symbol(format!("{}@{}", s.0, &self.step.to_string())))
        } else if self.next_variables.contains_key(&s.0) {
            let next = self.step + 1;
            let current_variable_name = self.next_variables.get(&s.0).unwrap();
            Ok(Symbol(format!(
                "{}@{}",
                current_variable_name,
                &next.to_string()
            )))
        } else {
            Ok(s)
        }
    }
}