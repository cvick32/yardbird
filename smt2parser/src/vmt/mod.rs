use std::collections::HashMap;

use action::Action;
use array_abstractor::ArrayAbstractor;
use bmc::BMCBuilder;
use itertools::Itertools;
use smt::SMTProblem;
use utils::{get_transition_system_component, get_variables_and_actions};
use variable::Variable;

use crate::concrete::{Command, FunctionDec, Identifier, Sort, Symbol, SyntaxBuilder, Term};

static PROPERTY_ATTRIBUTE: &str = "invar-property";
static TRANSITION_ATTRIBUTE: &str = "trans";
static INITIAL_ATTRIBUTE: &str = "init";

mod action;
mod array_abstractor;
mod bmc;
mod smt;
mod utils;
mod variable;

/// VMTModel represents a transition system given in VMT format.
/// The VMT specification is no longer available but there is an example here:
/// https://es-static.fbk.eu/people/griggio/ic3ia/
#[derive(Clone, Debug)]
pub struct VMTModel {
    sorts: Vec<Command>,
    state_variables: Vec<Variable>,
    function_definitions: Vec<Command>,
    actions: Vec<Action>,
    initial_condition: Term,
    transition_condition: Term,
    property_condition: Term,
}

#[derive(Debug)]
pub enum VMTError {
    UnknownCommand(String),
}

impl VMTModel {
    pub fn checked_from(commands: Vec<Command>) -> Result<Self, VMTError> {
        let number_of_commands = commands.len();
        assert!(number_of_commands > 3, "Not enough commands for VMT model!");
        let property_condition: Term =
            get_transition_system_component(&commands[number_of_commands - 1], PROPERTY_ATTRIBUTE);
        let transition_condition: Term = get_transition_system_component(
            &commands[number_of_commands - 2],
            TRANSITION_ATTRIBUTE,
        );
        let initial_condition: Term =
            get_transition_system_component(&commands[number_of_commands - 3], INITIAL_ATTRIBUTE);
        let mut variable_commands: HashMap<String, Command> = HashMap::new();
        let mut sorts: Vec<Command> = vec![];
        let mut variable_relationships = vec![];
        let mut function_definitions = vec![];
        for (i, command) in commands.iter().enumerate() {
            if i < number_of_commands - 3 {
                // Check whether a variable should be action, state, or local
                match command {
                    Command::DeclareFun {
                        symbol,
                        parameters,
                        sort: _,
                    } => {
                        if parameters.is_empty() {
                            variable_commands.insert(symbol.0.clone(), command.clone());
                        } else {
                            function_definitions.push(command.clone());
                        }
                    }
                    Command::DefineFun { sig: _, term: _ } => {
                        variable_relationships.push(command);
                    }
                    Command::DeclareSort {
                        symbol: _,
                        arity: _,
                    } => {
                        sorts.push(command.clone());
                    }
                    _ => return Err(VMTError::UnknownCommand(command.to_string())),
                }
            }
        }
        let (state_variables, actions) =
            get_variables_and_actions(variable_relationships, variable_commands);

        Ok(VMTModel {
            sorts,
            function_definitions,
            state_variables,
            actions,
            initial_condition,
            transition_condition,
            property_condition,
        })
    }

    /// Clones the current model, rewrites all usages of Arrays into uninterpreted functions
    /// and returns the abstracted VMTModel.
    pub fn abstract_array_theory(&self) -> VMTModel {
        let mut array_types: HashMap<String, String> = HashMap::new();
        array_types.insert("Int".to_string(), "Int".to_string());
        let mut abstractor = ArrayAbstractor {
            visitor: SyntaxBuilder,
            array_types,
        };
        let mut abstracted_commands = vec![];
        for command in self.as_commands() {
            abstracted_commands.push(command.accept(&mut abstractor).unwrap());
        }
        let mut array_definitions = abstractor.get_array_type_definitions();
        array_definitions.extend(abstracted_commands);
        VMTModel::checked_from(array_definitions).unwrap()
    }

    pub fn unroll(&self, length: u8) -> SMTProblem {
        let mut builder = BMCBuilder {
            visitor: SyntaxBuilder,
            current_variables: self.get_all_current_variable_names(),
            next_variables: self.get_all_next_variable_names(),
            step: 0,
        };
        let mut smt_problem = SMTProblem::new(&self.sorts, &self.function_definitions);

        smt_problem.add_assertion(&self.initial_condition, builder.clone());
        for _ in 0..length {
            // Must add variable definitions for each variable at each time step.
            smt_problem.add_variable_definitions(
                &self.state_variables,
                &self.actions,
                builder.clone(),
            );
            smt_problem.add_assertion(&self.transition_condition, builder.clone());
            builder.add_step();
        }
        // Don't forget the variable definitions at time `length`.
        smt_problem.add_variable_definitions(&self.state_variables, &self.actions, builder.clone());
        smt_problem.add_property_assertion(&self.property_condition, builder.clone());
        assert!(
            smt_problem.init_and_trans_length() == (length + 1).into(),
            "Unrolling gives incorrect number of steps {} for length {}.",
            smt_problem.init_and_trans_length(),
            length
        );
        smt_problem
    }

    pub fn as_commands(&self) -> Vec<Command> {
        let mut commands = self.sorts.clone();
        commands.extend(self.function_definitions.clone());
        for variable in self.state_variables.clone() {
            commands.extend(variable.as_commands());
        }
        for action in self.actions.clone() {
            commands.extend(action.as_commands());
        }
        let init_command = Command::DefineFun {
            sig: FunctionDec {
                name: Symbol("init".to_string()),
                parameters: vec![],
                result: Sort::Simple {
                    identifier: Identifier::Simple {
                        symbol: Symbol("Bool".to_string()),
                    },
                },
            },
            term: self.initial_condition.clone(),
        };
        commands.push(init_command);
        let trans_command = Command::DefineFun {
            sig: FunctionDec {
                name: Symbol("trans".to_string()),
                parameters: vec![],
                result: Sort::Simple {
                    identifier: Identifier::Simple {
                        symbol: Symbol("Bool".to_string()),
                    },
                },
            },
            term: self.transition_condition.clone(),
        };
        commands.push(trans_command);
        let prop_command = Command::DefineFun {
            sig: FunctionDec {
                name: Symbol("prop".to_string()),
                parameters: vec![],
                result: Sort::Simple {
                    identifier: Identifier::Simple {
                        symbol: Symbol("Bool".to_string()),
                    },
                },
            },
            term: self.property_condition.clone(),
        };
        commands.push(prop_command);

        commands
    }

    pub fn print_stats(&self) {
        println!("Number of Variables: {}", self.state_variables.len());
        println!("Number of Actions: {}", self.actions.len());
        println!("Number of Sorts: {}", self.sorts.len());
    }

    pub fn as_vmt_string(&self) -> String {
        self.as_commands()
            .iter()
            .map(|command| format!("{}", command.clone().accept(&mut SyntaxBuilder).unwrap()))
            .join("\n")
    }

    fn get_all_current_variable_names(&self) -> Vec<String> {
        let mut state_variable_names: Vec<String> = self
            .state_variables
            .iter()
            .map(|var| var.get_current_variable_name().clone())
            .collect();
        let mut action_names: Vec<String> = self
            .actions
            .iter()
            .map(|action| action.get_current_action_name().clone())
            .collect();
        state_variable_names.append(&mut action_names);
        state_variable_names
    }

    fn get_all_next_variable_names(&self) -> HashMap<String, String> {
        self.state_variables
            .iter()
            .map(|var| {
                (
                    var.get_next_variable_name().clone(),
                    var.get_current_variable_name().clone(),
                )
            })
            .collect()
    }
}
