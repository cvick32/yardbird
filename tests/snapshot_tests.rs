use insta::assert_debug_snapshot;
use smt2parser::vmt::VMTModel;
use std::{
    fs::read_dir,
    path::Path,
    sync::mpsc::{self, RecvTimeoutError},
    thread,
    time::Duration,
};
use yardbird::proof_loop;

#[derive(Debug)]
enum BenchStatus {
    Good,
    Timeout,
    Panic,
}

fn run_with_timeout<F, T>(f: F, timeout: Duration) -> (BenchStatus, T)
where
    F: FnOnce() -> T + Send + 'static,
    T: Send + Default + 'static,
{
    let (tx, rx) = mpsc::channel();
    let _ = thread::spawn(move || {
        let result = f();
        if let Ok(()) = tx.send(result) {}
    });

    match rx.recv_timeout(timeout) {
        Ok(insts) => (BenchStatus::Good, insts),
        Err(RecvTimeoutError::Timeout) => (BenchStatus::Timeout, T::default()),
        Err(RecvTimeoutError::Disconnected) => (BenchStatus::Panic, T::default()),
    }
}

#[allow(unused)]
#[derive(Debug)]
struct BenchmarkResult {
    example_name: String,
    status: BenchStatus,
    used_instantiations: Vec<String>,
}

fn run_benchmark(filename: impl AsRef<Path>) -> BenchmarkResult {
    let conrete_model = VMTModel::from_path(filename.as_ref()).unwrap();
    let mut abstract_model = conrete_model.abstract_array_theory();
    let (status, used_instantiations) = run_with_timeout(
        move || {
            let mut used_instantiations = vec![];
            proof_loop(&10_u8, &mut abstract_model, &mut used_instantiations).unwrap();
            used_instantiations
        },
        Duration::from_secs(20),
    );
    BenchmarkResult {
        example_name: filename.as_ref().to_string_lossy().to_string(),
        status,
        used_instantiations,
    }
}

#[test]
fn test_examples() {
    for entry in read_dir("./examples/").unwrap() {
        let path = entry.unwrap().path();
        let name = path.file_name().unwrap().to_str().unwrap();
        assert_debug_snapshot!(name, run_benchmark(&path));
    }
}
