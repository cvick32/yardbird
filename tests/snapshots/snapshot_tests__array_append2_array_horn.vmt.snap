---
source: tests/snapshot_tests.rs
expression: run_benchmark(&path)
---
BenchmarkResult {
    example_name: "./examples/array_append2_array_horn.vmt",
    status: Good,
    used_instantiations: [
        "(= (Read-Int-Int (Write-Int-Int A K W) K) W)",
        "(= (Read-Int-Int (Write-Int-Int Array-Int-Int!val!0 N W) N) W)",
        "(= (Read-Int-Int (Write-Int-Int A K W) K) (Read-Int-Int A K))",
        "(= (Read-Int-Int (Write-Int-Int Array-Int-Int!val!0 N W) N) (Read-Int-Int Array-Int-Int!val!0 N))",
    ],
}
